proxy/api/src/notification.rs

```diff
@@ -0,0 +1,47 @@
+//! Machinery to signal significant events to clients.
+
+use std::collections::HashMap;
+use std::net::SocketAddr;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::Arc;
+
+use serde::Serialize;
+use tokio::sync::mpsc;
+use tokio::sync::RwLock;
+
+/// Significant events happening during proxy runtime.
+#[derive(Clone, Debug, Serialize)]
+pub enum Notification {
+    /// Our local peer started listening on a local socket.
+    LocalPeerListening(SocketAddr),
```

Comment:
I find the "local peer" terminology kind of confusing in this context.. Maybe just `Listening`? :man_shrugging:

Reply:
The reason I prefixed it overly verbose is to account for potentiall other things we start to listen on a socket, we already listen for http requests on another addr, so `Listening` can be qutie ambiguous.

Reply:
What is this event supposed to mean to the frontend? Could something like `NetworkReady` work?

Reply:
To any api consumer it will signal that the peer we run locally as part of the proxy has physically been bound to a socket on the local machine and is listening on that. Examples of other notifications are: `PeerConnected`, `PeerDisconnected` - which refer to remote peers connecting to our local peer. Those are at least the terms I'm thinking of.

I see where you going with `NetworkReady` as it is trying to convey things on a higher level. But could also mean different things, maybe we consider readiness once we have connected to a couple of peers and have synced up. Not necessarily what I'm advocated for, my point is we would need to define what readiness meanss

Reply:
How is this event generated? Doesn't our app fail if the `Peer` doesn't `accept` to get the `PeerApi`?

Reply:
It's currently not generated yet. We will hook this up wherever we end up making sense of protocol events, currently:

https://github.com/radicle-dev/radicle-upstream/blob/9f8981de6fc9b00f2d564b322a48f0050ab5543b/proxy/coco/src/peer.rs#L63-L67

This will be one of the follow-ups to surface these to consumers. One of the use cases I have in mind at the moment, is a debug/developer view where you can see the log of all events from the proxy in the UI, to help understand what is going on and/or you don't have access to the logs.

Reply:
Would it not make more sense to do:
```rust
pub enum Notification {
  Protocol(ProtocolEvent),
}
```
using [`ProtocolEvent`](https://github.com/radicle-dev/radicle-link/blob/master/librad/src/net/protocol.rs#L56-L61) as the variant?

Reply:
I don't think it's wise to just pass through the `librad` protocol events and not make an active effort to curate what notifications we are emitting with what semantics.

Reply:
The curation can still be done by `match`ing on `Protocol`. On the flip side, not using `ProtocolEvent` means we have to maintain a mapping.

Reply:
My point is we want to maintain that mapping and `notification.rs` is the place to do that as this is where we curate the list of notifications the proxy can emit and will surface to clients. Let's take the `ProtocolEvent::Gossip` variant, a client needs to have deep knowledge of the protocol library in order to make it palatable for a user. Instead the notifications should be self-contained and ready to be presented. A use-case I'm thinking along is a`tail` of notifications showing us peer connection events, announcements and requests of heads to help understand network activity.

Reply:
Ok, but to expand what I mean by "The curation can still be done by `match`ing" is that we could have something like:
```rust
enum Message {
  Interesting { event_name: String, value: Json }
  Uninteresting,
}

impl Message {
  fn fire(&self, subscription: Subscription) {
    match self {
      Uninteresting => { /* do nothing */ }
      Interesting { event_name, value } => /* fire message off to Subscription */
    }
  }
}

enum Notification {
  Protocol(ProtocolEvent)
}

impl Notifiction {
  fn curate(&self) -> Message {
    match self {
      Protocol(protocol) => if protocol.is_gossip() { Uninteresting } else { make_interesting(protocol) }
    }
  }
}
```

A bit pseudo-codey but hopefully it gets my point across :)

Reply:
Where would you expect this to live, i.e. at which layer? Don't really see why we would need this extra machinery if we can already determine interesting events with the definition of the `Notification` variants.

Reply:
I could see both of these living here. The idea being that notification is rich with information, and this gets crushed down to something that's easy to communicate, i.e. this `Message` thing. Then we have the ability to transform or ignore the rich information into the easy to communicate information.

The advantage being that once we realise we're interested in having gossip notifications, we just change the handling of that case and turning into a message. Rather than adding a new case and adding handling logic for each place that case pops up.

Another advantage being that we don't need to come up with new names, which prompted this thought in the first place :smile: